cluster.go

```diff
@@ -1,90 +1,142 @@
+	serverStateLock := new(sync.Mutex)
+
+	/* Election Timer: Checks if timeout is surpassed and starts election. Timeout is reached when:
+	 * 1. no message from leader or
+	 * 2. when election took too long (e.g. due to tie / no leader elected)
+	 */
+	go func () {
+		for {
+			timeElapsed := time.Now().Sub(timeSinceLastUpdate)
+			if timeElapsed.Milliseconds() > ElectionTimeOut {
+				isElection = true
+				timeSinceLastUpdate = time.Now()
+				go elect(&state, voteChannels, leaderCommunicationChannels)
+			}
+			time.Sleep(electionThreadSleepTime)
+		}
+	}()
+
+	// receive messages from leader
+	go func () {
+		for newLogEntry := range leaderCommunicationChannels[state.ServerId] {
+			if isElection { //received message from leader during election,
+				serverStateLock.Lock()
+				state.Role = FollowerRole // for candidates that lost the election
+				serverStateLock.Unlock()
+				isElection = false
+			}
+			timeSinceLastUpdate = time.Now()
+			fmt.Print("New log entry received from leader:", newLogEntry, "\n")
+			//process log entry here
+		}
+		done <- true
+	}()
 }
 
-
-func elect(state ServerState, voteChannels *[ClusterSize]chan Vote, electionTime *time.Timer) {
-	startTimeOut := rand.Intn(150) + 150
-
-	startTime := time.NewTimer(time.Duration(startTimeOut) * time.Millisecond)
+/* Begins an election and handles the following events:
+ * 1. Election timeout reaches threshold -> Become candidate
+ * 2. External Request to vote -> Give vote, ignore if voted already
+ */
+func elect(
+	state * ServerState,
+	voteChannels *[ClusterSize]chan Vote,
+	leaderCommunicationChannels *[ClusterSize]chan LogEntry,
+	) {
+	timeUntilElectionStart := rand.Intn(150) + 150
+	electionStartTimer := time.NewTimer(time.Duration(timeUntilElectionStart) * time.Millisecond)
+	serverStateLock := new(sync.Mutex)
 
 	select {
-	case <-startTime.C: // candidate
+	case <-electionStartTimer.C: // candidate
 		fmt.Println("Server ", state.ServerId, " is a candidate")
-		// start election
-		state.CurrentTerm++
 
-		// vote for self
-		state.VotedFor = state.ServerId
-		
-		// reset election timer
-		electionTime.Reset(time.Duration(ElectionTimeOut) * time.Millisecond)
+		// lock while transitioning to candidate
+		serverStateLock.Lock()
+		state.CurrentTerm++
+		state.Role = CandidateRole
+		state.VotedFor = state.ServerId // vote for self
+		serverStateLock.Unlock()
 		
 		//count votes
 		winnerChannel := make(chan bool)
 		go requestVotes(state, voteChannels, winnerChannel)
 		
 		select {
 		case <-winnerChannel: // got enough votes
-			// start sending heartbeats (blank appendentries)
-		case <-electionTime.C: // election timed out
-			// restart election
+			for serverIndex, leaderCommunicationChannel := range leaderCommunicationChannels {
+				leaderCommunicationChannel := leaderCommunicationChannel
+				go func () {
+					leaderCommunicationChannel <- LogEntry{serverIndex, state.CurrentTerm, KeyValue{"assert", "dominance"}}
```

Comment:
Lmao love the "assert dominance" key value pair

Reply:
oh shoot hahaha I forget to change this back to an empty Key-Value because I think this could mess us up later lmao